scmrepo/git/backend/dulwich/__init__.py

```diff
@@ -597,16 +597,34 @@ def describe(
         match: Optional[str] = None,
         exclude: Optional[str] = None,
     ) -> Optional[str]:
+        from dulwich.porcelain import active_branch
+
         if not base:
             base = "refs/tags"
-        for ref in self.iter_refs(base=base):
+
+        if base == "refs/heads":
+            try:
+                current_branch = (
+                    base + "/" + os.fsdecode(active_branch(self.repo))
+                )
+            except (IndexError, ValueError):
+                current_branch = None
+
+        best_ref = None
+
+        for ref in sorted(self.iter_refs(base=base)):
```

Comment:
So, one problem with this implementation is that it sorting will be slow in the event that a repo has a lot of refs. `git-describe` doesn't really guarantee any particular ordering - from the documentation it just has to return any ref (with the appropriate base) that matches the requested revision.

> git-describe - Give an object a human readable name based on an available ref

The way that we are setting `best_ref` also requires that we finish the entire iteration before returning a result (so we are even slower in that we sort the refs, and also have to re-iterate over the entire sorted list to find `best_ref`).

Preferring the active branch is fine when `base=refs/heads`, but it would be better for us to just check if the current `HEAD` ref is pointing to a branch, and if so, immediately check if that branch points to `rev` before doing any iteration at all.

Otherwise, we can just fall back to doing the existing unsorted iteration and return the first matching result we find.

Reply:
>  Otherwise, we can just fall back to doing the existing unsorted iteration and return the first matching result we find.

So, OK to give a random branch name if we are using for example `dvc exp show -A` for a current not checkouted one?

Reply:
In my opinion, yes. It's still up for discussion in the original issue, but in general it's a low priority UI issue.

The active branch/HEAD case is worth addressing specifically because by default we only show HEAD, but in the general case for other revs I would prioritize keeping the current performance over making any further behavior changes to `describe`.

(see also: https://github.com/iterative/dvc/issues/6051#issuecomment-847820247)